runtime/cudaq/builder/kernel_builder.cpp

```diff
@@ -156,31 +156,94 @@ bool isArgStdVec(std::vector<QuakeValue> &args, std::size_t idx) {
   return args[idx].isStdVec();
 }
 
+/// @brief Search the given `FuncOp` for all `CallOps` recursively.
+/// If found, see if the called function is in the current `ModuleOp`
+/// for this `kernel_builder`, if so do nothing. If it is not found,
+/// then find it in the other `ModuleOp`, clone it, and add it to this
+/// `ModuleOp`.
+void addAllCalledFunctionRecursively(
+    func::FuncOp &function, ModuleOp &currentModule,
+    mlir::OwningOpRef<mlir::ModuleOp> &otherModule) {
+
+  std::function<void(func::FuncOp func)> visitAllCallOps;
+  visitAllCallOps = [&](func::FuncOp func) {
+    func.walk([&](func::CallOp callOp) {
+      // Don't add if we already have it
+      if (currentModule.lookupSymbol<func::FuncOp>(callOp.getCallee()))
```

Comment:
Do we want to check for Quake ApplyOp cases here?

Reply:
Yea I think so. Good catch.

Reply:
Updated. Please check it out.